app/models/manageiq/providers/amazon/agent_coordinator_worker/runner.rb

```diff
@@ -0,0 +1,63 @@
+class ManageIQ::Providers::Amazon::AgentCoordinatorWorker::Runner < MiqWorker::Runner
+  def do_before_work_loop
+    @coordinators = self.class.all_agent_coordinators_in_zone
+  end
+
+  def do_work
+    @coordinators.each do |m|
+      alive_agents = m.alive_agent_ids
+
+      _log.info("Alive agents in EMS(guid=#{m.ems.guid}): #{alive_agents}.")
+
+      # Only setup agents when:
+      # 1. there is no running agents;
+      # 2. get new requests;
+      # 3. coordinator is not in deploying agent state;
+      m.startup_agent if alive_agents.empty? && !m.request_queue_empty? && !m.deploying?
+
+      # Turn flag off if deploying is done.
+      m.deploying = false unless alive_agents.empty?
+    end
+
+    # Amazon providers may be added/removed. Keep monitor and update if needed.
+    latest_ems_guids = self.class.amazon_ems_guids
+    coordinator_guids = @coordinators.collect { |m| m.ems.guid }
+    self.class.refresh_coordinators(@coordinators, coordinator_guids, latest_ems_guids)
+  end
+
+  def self.agent_coordinator_by_guid(guid)
+    all_agent_coordinators_in_zone.find { |e| e.ems.guid == guid }
+  end
+
+  def self.ems_by_guid(guid)
+    all_valid_ems_in_zone.detect { |e| e.guid == guid }
+  end
+
+  def self.amazon_ems_guids
+    all_amazon_ems_in_zone.collect(&:guid)
+  end
+
+  def self.all_ems_in_zone
+    ExtManagementSystem.where(:zone_id => MiqServer.my_server.zone.id).to_a
```

Comment:
MiqServer.my_server.zone.ext_management_systems

Reply:
This is also already part of the PerEmsTypeWorkerMixin, so this is all duplication.